Sources/ArgumentParser/Usage/ToolInfo.swift

```diff
@@ -0,0 +1,139 @@
+//===----------------------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import Foundation
+
+fileprivate extension Collection {
+  var `optional`: Self? { isEmpty ? nil : self }
+}
+
+@_spi(ToolInfo)
+public struct ToolInfoHeader: Decodable {
+  public var serializationVersion: Int
+
+  public init(serializationVersion: Int) {
+    self.serializationVersion = serializationVersion
+  }
+}
+
+@_spi(ToolInfo)
+public struct ToolInfoV0: Codable, Hashable {
+  public var serializationVersion: Int = 0
+  public var command: CommandInfoV0
+
+  public init(
+    serializationVersion: Int = 0,
+    command: CommandInfoV0
+  ) {
+    self.serializationVersion = serializationVersion
```

Comment:
What's your plan for versioning this? Can `ToolInfoV0` support serialization version 1?

Reply:
My initial thought is that the serialization version would essentially be a semver major indicator used to present better error messages.

Evolving the serialized format could follow these rules:
- New fields can be added freely
- Many fields can be omitted removed (because they are optional)
- Changes to the encoded format can be handled by manually implementing `init(from decoder: Decoder)`

If a backwards incompatible change needs to be made, we can bump the serialization version which would have a more legible error than "decoding failure".

(However I'm not too happy with this plan, so I'd be really happy to get suggestions)